test/e2e/e2e_suite_test.go

```diff
@@ -0,0 +1,207 @@
+	useExistingCluster bool
+
+	// e2eConfig to be used for this test, read from configPath.
+	e2eConfig *Config
+
+	// artifactFolder is the folder to store e2e test artifacts.
+	artifactFolder string
+
+	// skipCleanup prevents cleanup of test resources e.g. for debug purposes.
+	skipCleanup bool
+
+	// clusterProxy allows to interact with the cluster to be used for the e2e tests.
+	clusterProxy framework.ClusterProxy
+
+	// clusterProvider manages provisioning of the cluster to be used for the e2e tests.
+	// Please note that provisioning will be skipped if e2e.use-existing-cluster is provided.
+	clusterProvider bootstrap.ClusterProvider
+)
+
+func init() {
+	flag.StringVar(&configPath, "e2e.config", "", "path to the e2e config file")
+	flag.StringVar(&artifactFolder, "e2e.artifacts-folder", "_artifacts", "folder where e2e test artifact should be stored")
+	flag.BoolVar(&useExistingCluster, "e2e.use-existing-cluster", false, "if true, the test uses the current cluster instead of creating a new one (default discovery rules apply)")
+	flag.BoolVar(&skipCleanup, "e2e.skip-resource-cleanup", false, "if true, the resource cleanup after tests will be skipped")
+}
+
+func TestE2e(t *testing.T) {
+	g := NewWithT(t)
+	ctrl.SetLogger(klog.Background())
+
+	// ensure the artifacts folder exists
+	g.Expect(os.MkdirAll(artifactFolder, 0755)).To(Succeed(), "Invalid test suite argument. Can't create e2e.artifacts-folder %q", artifactFolder) //nolint:gosec
+
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "E2e Suite")
+}
+
+var _ = SynchronizedBeforeSuite(func() []byte {
+	var kubeconfigPath string
+	Expect(configPath).To(BeAnExistingFile(), "Invalid test suite argument. e2e.config should be an existing file.")
+	e2eConfig = LoadE2EConfig(configPath)
+
+	if useExistingCluster {
+		kubeconfigPath = os.Getenv("KUBECONFIG")
+		if kubeconfigPath == "" {
+			kubeconfigPath = os.Getenv("HOME") + "/.kube/config"
+		}
+	} else {
+		clusterProvider = bootstrap.CreateKindBootstrapClusterAndLoadImages(ctx, bootstrap.CreateKindBootstrapClusterAndLoadImagesInput{
+			Name:   "bmo-e2e",
+			Images: e2eConfig.Images,
+		})
+		Expect(clusterProvider).ToNot(BeNil(), "Failed to create a cluster")
+		kubeconfigPath = clusterProvider.GetKubeconfigPath()
+	}
+	Expect(kubeconfigPath).To(BeAnExistingFile(), "Failed to get the kubeconfig file for the cluster")
+
+	scheme := runtime.NewScheme()
+	framework.TryAddDefaultSchemes(scheme)
+	clusterProxy := framework.NewClusterProxy("bmo-e2e", kubeconfigPath, scheme)
+	Expect(clusterProxy).ToNot(BeNil(), "Failed to get a cluster proxy")
+
+	DeferCleanup(func() {
+		clusterProxy.Dispose(ctx)
+	})
+
+	os.Setenv("KUBECONFIG", clusterProxy.GetKubeconfigPath())
+
+	if e2eConfig.GetVariable("DEPLOY_CERT_MANAGER") != "false" {
+		// Install cert-manager
+		By("Installing cert-manager")
+		cmd := exec.Command("cmctl", "x", "install")
```

Comment:
is there any advantage of using `cmctl` instead of just using `kubectl` if so I guess we should unify it with capm3 as well https://github.com/metal3-io/cluster-api-provider-metal3/blob/5ae544453b6e8cc4bd4f6670665a10d8066e8f5f/test/e2e/upgrade_clusterctl_test.go#L194

Reply:
See my answer to Kashifs comment above